Exiles on 1.1

anselor opened the pull request:
Aggregation of many PRs

Integrates the following PRs

#123 
#136 
#138 
#145 
#148 
#150 
#155 

#157 


Supersedes #139

billyjbryant committed:
Updating GHA to push  docker images tagged 'edge/upcoming' from merges to upcoming/*

mtoner23 committed:
Changing mqtt padding from zeros to bytes

mtoner23 committed:
remove .vscode from git ignore

mtoner23 committed:
Fix padding in mqttapi::make_mqtt_pkt

snoj committed:
wss:// support

Added ability to use wss for ssl deployments (eg traefik/nginx proxy manager).

billyjbryant committed:
1.0.1 - Fix mqtt bug release

treitmayr committed:
Add initial support for the AnkerMake M5C

This change allows to monitor and send MQTT messages from/to the
AnkerMake M5C on the command line as well as observe the printer status
via the Web GUI.
It is also possible to send print jobs from your slicer to the printer
without using AnkerMake Slicer or Studio (see README.md).
Due to the M5's additional hardware (camera, light) still being expected
to exist, the Web GUI will continuously wait for a camera connection.
This should be fixed in the future.

treitmayr committed:
Add support for fetching configuration via login

Implemented new command 'config login', which also supports
captcha queries, should they happen.

treitmayr committed:
Upon 'config login' try to rescue already configured printe IPs

If there is already a valid ankerctl configuration and the user
just wants to re-login via 'config login' then the printer IP
addresses are restored to their original value if the server did
not provide those IP addresses (for me it never does).
Also, moved around some code parts.

treitmayr committed:
Add possiblity to log in via web interface

This new form also supports captcha requests.
Also fixed is the behavior that loading/importing a new configuration
led to the printer IP addresses being cleared.

treitmayr committed:
Support updating printer IP addresses from CLI

For this the flag '--store' was added to 'pppp lan-search'.

treitmayr committed:
Support updating printer IP addresses via web interface

treitmayr committed:
Optimize chunk splitting in gcode file transfers

Instead of copying the whole gcode file data and then splitting it into
32 KB chunks (2nd copy), use a generator function which emits the chunks
along with the current offset to be used in the AABB request.

treitmayr committed:
Split PPPP badge into proper PPPP and VIDEO badges

Now the PPPP badge just shows the status of the PPPP connection, and
this also works for the AnkerMake M5C. The new VIDEO badge now has the
same behavior as the original PPPP badge in that it shows the status of
the video stream.

treitmayr committed:
Bind socket on broadcasts when running in Windows OS

This fixes issue #152.

treitmayr committed:
Provide more privacy by redacting the country code

This affects both the CLI 'config show' as well as the 'Setup' tab of
the web interface.

treitmayr committed:
Show proper error messages when printer IP address is not available

anselor committed:
Changed docker compose file to accept optional configuration parameters. Added helper shell script for docker.
Changed docker import to accept a login.json in the currect directory if available.

anselor committed:
Changed default yaml for exiles to pull pre-built exiles image
Added stand-alone dockerfile for building runtime container.

anselor committed:
Changed ankerctl to read from .env or .flaskenv on startup.
Added pyinstaller packaging for windows users.
Updated requirements needed to run transwarp.

sondregronas commented on static/ankersrv.js:
```diff
@@ -355,18 +426,31 @@ $(function () {
 
+    sockets.pppp_state = new AutoWebSocket({
+        name: "PPPP socket",
+        url: `ws://${location.host}/ws/pppp-state`,
```
Found an error in the pppp_state that would prevent PPPP connection over https:

```suggestion
        url: `${location.protocol.replace('http','ws')}${location.host}/ws/pppp-state`,
```

anselor replied:
@treitmayr This was from your changes in this PR:
https://github.com/Ankermgmt/ankermake-m5-protocol/pull/145/commits/1089b493d13e6c3b6abbe4a9168972407702e8c4

treitmayr replied:
My changes were done in a consistent way in the existing code base at the time of the commit. However, the merge of [d6c727b](https://github.com/Ankermgmt/ankermake-m5-protocol/pull/160/commits/d6c727b378140d58395d50f12482fec53864b1d3) introduced wss support, again on the code base from back then. This is the explanation for the current state.
However, I agree with @sondregronas to change the affected line in the proposed way.

sondregronas committed:
Update ankersrv.js

sondregronas committed:
Add some exceptions & shutdown on critical errors

sondregronas committed:
Forgot to import logging

sondregronas committed:
Remove logging from exceptions

Seems to not work well to have them separated.

This does now work, but it raises a lot more exceptions than just AnkerCriticalError (KeyboardInterrupt & TypeError) - probably from separate threads

sondregronas committed:
Rework auto-restart on pppp-disconnect

Restarting pppp will initiate new connection attempts. Will monitor this for a bit, a lot more graceful than a force shutdown

sondregronas committed:
Re-add deleted comments

Removed these by mistake

sondregronas committed:
Add timestamp to log - everything seems to work :smile:

Will leave this running for a while

sondregronas committed:
Add an /api/ankerctl/status endpoint

Felt fitting to add this here - though I can change target to the main repo

sondregronas committed:
Don't restart an already starting service

sondregronas committed:
More trial and error with pppp restarting

Hopefully the last commit - has survived 3 lost connections so far. (The previous method had the service always running, which was stable but not realistic)

sondregronas committed:
Some more effort into restarting PPPP - turns out restart is unreliable (by design)

sondregronas committed:
Working code; with a caveat

sondregronas committed:
Update API & update some dependencies

Haven't had a single error message since updating requirements!

sondregronas committed:
Fix webserver access without config

Without this, logging in was never an option (would crash on start)

sondregronas committed:
Addressing some minor oversights + adding version to api status

sondregronas committed:
Reduce log level of pppp connection attempts

sondregronas commented:
Just commenting here to say that I've opened a PR to this PR with some more changes - let me know if you want me to switch targets to the main repo. PR: https://github.com/anselor/ankermake-m5-protocol/pull/4

### Changes within that PR to this one

- updating `pycryptodomex`, `flask`, `flask-sock` to the latest version + adding `python-dotenv` to dependencies
- addresses the change mentioned above in `ankersrv.js`
- restart the pppp api worker (by calling `worker_start()`, which restarts the api state) on pppp failure (see notes below)
- start the videoqueue service on `/video` if it isn't running, allowing webrtc access via go2rtc (see #162)
- bugfix: allow webserver access with missing credentials (webserver couldn't start without any camera present in cfg)
- `/api/ankerctl/status` endpoint, displays a JSON of every services state and the `app_api_version` dict

### Note regarding pppp:

With this version, pppp should not crash (at least not as often). I've noticed before it would stop working after 2-3 hours of uptime, (but that could also be my really old xeon failing). Instead of crashing it will "flicker" for lack of better words, where sending gcode will _sometimes_ work, and the videostream will have corrupted B/P-frames that only get worse over time. Reloading services are to no avail. Restarting the entire application every 2 hours does circumvent this as far as I can tell.

I tried getting the webserver to run with a different WSGI server, but had no luck. I only tested `gevent` or `eventlet`. I wanted to try gunicorn but from my understanding it would require some refactoring to how the cfg is loaded so I stopped trying.

### All of these prs combined _should_ resolve (at least to my limited understanding):

#163, #161, #158, partially #156 (PPPP will _work_ but stops being reliable), #154, #151, #147, #127 (might be unrelated?), #112

Might be missing a few issues as well that I haven't encountered myself.
